### PR TITLE
issue/4718 - Introduce edd_get_payment() and caching for payments

### PR DESCRIPTION
**Pull request by sunnyratilal**
#4718

### CODE REVIEW

File: includes/payments/functions.php [hunk 1]
```diff
@@ -579,6 +579,34 @@ function edd_check_for_existing_payment( $payment_id ) {
...
 
 /**
+ * Retrieves an instance of EDD_Payment for a specified ID.
+ *
+ * @since 2.7
+ *
+ * @param int $id Payment ID.
+ * @return mixed false|object EDD_Payment if a valid payment ID, false otherwise.
+ */
+function edd_get_payment( $payment_id = 0 ) {
```
Comment (pippinsplugins) [inline includes/payments/functions.php]: This should also support passing in a `WP_Post` object and an `EDD_Payment` object.

**Comment by cklosowski** [comment]
Also, a transaction ID, since EDD_Payment supports that currently in the
construct.

Chris Klosowski
cklosowski@gmail.com
@cklosowski

**Comment by pippinsplugins** [comment]
Ah yes, txn :+1:

File: includes/payments/functions.php [hunk 2]
```diff
@@ -609,6 +609,45 @@ function edd_check_for_existing_payment( $payment_id ) {
...
+	} else {
+		$payment_id = $payment;
+	}
+
+	if ( empty( $payment_id ) ) {
+		return false;
+	}
+
+	$cache_key = md5( 'edd_payment' . $payment_id );
+	$_payment = wp_cache_get( $cache_key, 'payments' );
```
Comment (pippinsplugins) [inline includes/payments/functions.php]: `$_payment` is never returned when retrieved from cache due to the variable names being different.

File: includes/payments/functions.php [hunk 3]
```diff
@@ -13,6 +13,50 @@
...
+ * @param mixed int|EDD_Payment|WP_Post $payment Payment ID, EDD_Payment object or WP_Post object.
+ * @param bool                          $by_txn  Is the ID supplied as the first parameter
+ * @return mixed false|object EDD_Payment if a valid payment ID, false otherwise.
+ */
+function edd_get_payment( $payment_or_txn_id = null, $by_txn = false ) {
+	global $wpdb;
+
+	if ( is_a( $payment_or_txn_id, 'WP_Post' ) || is_a( $payment_or_txn_id, 'EDD_Payment' ) ) {
+		$payment_id = $payment_or_txn_id->ID;
+	} elseif ( $by_txn ) {
```
Comment (pippinsplugins) [inline includes/payments/functions.php]: This needs to check if the transaction ID is empty and bail if so.

Comment (sunnyratilal) [inline includes/payments/functions.php]: Fixed in b36b2b33d280ea87d24a772ee97653cb2b44bc26

**Pull request merged**